### PR TITLE
[Resolve #683] Creating a Diff command (Prototype)

### LINKED ISSUE #683: Enhancement: Show diff (and change set details) before updating a stack
# Current Behavior

Today, when a user runs a command to update a stack, they see a prompt like this:

```
$ sceptre update prod/vpc
Do you want to update 'prod/vpc' [y/N]:
```

While the user's goal might be to get `prod/vpc` into some new state, this isn't really enough information for her to confidently answer in the affirmative. There are many reasons the answer could be "N," even when trying to make a change:

1. The user hasn't checked out the latest version of the repo containing the sceptre project, and updating will actually revert older changes rather than apply new ones.
2. The user has the latest changes but forgot to add some resource or parameter.
3. The update is destructive or will cause the replacement of some resource, and the user wants to announce or further plan around such a change.
4. ...

# Desired Behavior

For every stack being updated (either a single stack or each stack in a stack group), show the diff of the template, the stack parameters, and if change sets are used, the change set details, before showing a stack update prompt.

For example:
```
$ sceptre update prod/vpc
Changes for 'prod/vpc/vpc.yaml':
Template:
  @@ -27,6 +27,8 @@
         Tags:
         - Key: Name
           Value: !Ref Name
  +      - Key: Test
  +        Value: '123'
     InternetGateway:
       Type: AWS::EC2::InternetGateway
       Properties:

Parameters:
  @@ -2,5 +2,5 @@
   Name: VPC
  -Tenancy: default
  +Tenancy: dedicated

Change Set:
Action Resource                   Replacement?
==============================================
Modify AttachGateway              False
Modify BastionSecurityGroup       Conditional
Modify NATSecurityGroup           Conditional
Modify PrivateSubnetSecurityGroup Conditional
Modify PublicRoute                Conditional
Modify PublicRouteTable           Conditional
Modify VPC                        Conditional

Do you want to update 'prod/vpc' with these changes? [y/N]:
```

If the general proposal sounds reasonable, we can discuss implementation details and any edge/corner cases we should consider.

### PR DESCRIPTION
**Pull request by jfalkenstein**
I'm posting this at this point for feedback. It's pretty important for my company's deployment process (and I suspect many others as well) to generate the "diff" on what this set of changes will do to the currently deployed group of stacks.

This creates what has been requested in a few different issues, most notably in #683 

The syntax here is basically `sceptre diff {command path}`

## What does this do, exactly?
This compares the CURRENTLY DEPLOYED stacks in the stack group with the current configuration of stacks and templates. 
It both compares configurations and generated templates, indicating the difference.

The heavy-lifting on this is done with the excellent DeepDiff library as well as the cfn-flip library that lets you flip between CF json and yaml.

The output can be directed to a file or dumped to stdout.

## What sort of output does this create?
********************************************************************************
```
--> Difference detected for stack sceptre-dev-jf-demo-bucket-website-http-bucket!
No config difference
--------------------------------------------------------------------------------
Template difference:
values_changed:
  root['Resources']['S3Bucket07682993']['Properties']['BucketName']:
    new_value: changed-environment-sceptre-demo
    old_value: dev-jf-sceptre-demo

********************************************************************************
********************************************************************************
--> Difference detected for stack sceptre-dev-jf-demo-demo-lambda-artifact-bucket!
This stack is not deployed yet!
--------------------------------------------------------------------------------
New config:
name: sceptre-dev-jf-demo-demo-lambda-artifact-bucket
notification_arns: []
parameters:
  ArtifactBucketName: dev-jf-random-artifacts
tags:
  Deployed_Using: Sceptre
  Eng_Business_Product: demo-lambda
  Environment: dev-jf
  Reason_For_Existance: To demonstrate Sceptre

--------------------------------------------------------------------------------
New template:
Parameters:
  ArtifactBucketName:
    Type: String
    Description: The name of the Amazon S3 bucket where uploaded SAM artifacts will
      be stored.
Resources:
  ArtifactBucket7410C9EF:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: !Ref 'ArtifactBucketName'
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: CdkStack/ArtifactBucket/Resource
  CDKMetadata:
    Type: AWS::CDK::Metadata
    Properties:
      Analytics: >-
        v2:deflate64:H4sIAAAAAAAAEyWNQQrCMBBFz+I+nTYK4lLsASz1BCEZIQ1NZDKji5C727Sr//g8/tegzwMMp7v55c660BebCKG82NigxnecDJkVGUmNKWYmsdzqp/BHdpoxJyGLjTfFefYpVtX28gXKQ2zAXTyoVhWTQ1hy/9VXuG3fS/a+I4nsV4T5yD/OizQKmAAAAA==
    Metadata:
      aws:cdk:path: CdkStack/CDKMetadata/Default
    Condition: CDKMetadataAvailable
Outputs:
  BucketName:
    Description: The ARN of the website bucket
    Value: !Ref 'ArtifactBucket7410C9EF'
Conditions:
  CDKMetadataAvailable: !Or
    - !Or
      - !Equals
        - !Ref 'AWS::Region'
        - af-south-1
      - !Equals
        - !Ref 'AWS::Region'
        - ap-east-1
      - !Equals
        - !Ref 'AWS::Region'
        - ap-northeast-1
      - !Equals
        - !Ref 'AWS::Region'
        - ap-northeast-2
      - !Equals
        - !Ref 'AWS::Region'
        - ap-south-1
      - !Equals
        - !Ref 'AWS::Region'
        - ap-southeast-1
      - !Equals
        - !Ref 'AWS::Region'
        - ap-southeast-2
      - !Equals
        - !Ref 'AWS::Region'
        - ca-central-1
      - !Equals
        - !Ref 'AWS::Region'
        - cn-north-1
      - !Equals
        - !Ref 'AWS::Region'
        - cn-northwest-1
    - !Or
      - !Equals
        - !Ref 'AWS::Region'
        - eu-central-1
      - !Equals
        - !Ref 'AWS::Region'
        - eu-north-1
      - !Equals
        - !Ref 'AWS::Region'
        - eu-south-1
      - !Equals
        - !Ref 'AWS::Region'
        - eu-west-1
      - !Equals
        - !Ref 'AWS::Region'
        - eu-west-2
      - !Equals
        - !Ref 'AWS::Region'
        - eu-west-3
      - !Equals
        - !Ref 'AWS::Region'
        - me-south-1
      - !Equals
        - !Ref 'AWS::Region'
        - sa-east-1
      - !Equals
        - !Ref 'AWS::Region'
        - us-east-1
      - !Equals
        - !Ref 'AWS::Region'
        - us-east-2
    - !Or
      - !Equals
        - !Ref 'AWS::Region'
        - us-west-1
      - !Equals
        - !Ref 'AWS::Region'
        - us-west-2

********************************************************************************
********************************************************************************
--> Difference detected for stack sceptre-dev-jf-demo-bucket-website-http-bucket-policy!
No config difference
--------------------------------------------------------------------------------
Template difference:
type_changes:
  root['Resources']['BucketPolicy']['Properties']['PolicyDocument']['Statement'][0]['Action']:
    new_type: list
    new_value:
      - s3:GetObject
      - s3:PutObject
    old_type: str
    old_value: s3:GetObject

********************************************************************************
No difference to deployed stack sceptre-dev-jf-demo-demo-lambda-demo-lambda

```

## PR Checklist

- [ ] Wrote a good commit message & description [see guide below].
- [ ] Commit message starts with `[Resolve #issue-number]`.
- [ ] Added/Updated unit tests.
- [ ] Added/Updated integration tests (if applicable).
- [ ] All unit tests (`make test`) are passing.
- [ ] Used the same coding conventions as the rest of the project.
- [ ] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [ ] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: sceptre/cli/diff.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+from sceptre.cli.helpers import catch_exceptions
+from sceptre.context import SceptreContext
+from sceptre.plan.plan import SceptrePlan
+from sceptre.plan.stack_differ import StackDiff, StackConfiguration
+
+
+@contextlib.contextmanager
+def nullcontext(enter_result=None):
+    yield enter_result
```
Comment (jfalkenstein) [inline sceptre/cli/diff.py:20]: Python 3.7 added this, but since we need to support 3.6 right now, I defined it here. Easy enough...

**Comment by alexharv074** [comment]
Please be aware that our site already implemented a `diff` command and I have raised a PR to show how we implemented it here https://github.com/Sceptre/sceptre/pull/1094

**Comment by alexharv074** [comment]
@jfalkenstein I had a play with this and I like it! I wonder if we should aim to combine the best aspects of both implementations. In my mind these are:

## Jon's branch

- :white_check_mark: The output is cool.
- :white_check_mark: Config is also diff'ed, more featureful etc.

## Alex's branch

- :white_check_mark: Allows selection of diff libraries.
- :white_check_mark: Defaults to unified diffs that people are often familiar with.

If possible, it would be nice, but not required, if backwards-compatibility could be maintained with our implementation. I can't immediately see any reason why not. It would just mean continue to support `-D/--differ` which defaults to `difflib`.

**Comment by jfalkenstein** [comment]
@alexharv074 You forgot to include that your branch has docs and tests lol. 

I think combining features would be best as well.

I don't see a fundamental issue with using different diffing libraries as yours does. I think the next step would be bringing over my StackDiffer class into your branch. (I have unit tests I wrote on an altered version of the StackDiffer on a tool for my company).  

Does unified diff just diff off strings? Or can it diff off dicts? I like using dicts for this because the order of keys doesn't matter. 

We should probably define a Diff interface and make a few implementations, which can receive two dicts and then return a "diff" result. We should probably care about the "output-format" arg, though.

**Comment by jfalkenstein** [comment]
Closing this in favor of #1113

**Pull request closed without merging**